R/select_nse.R

```diff
@@ -0,0 +1,512 @@
+.select_dollar <- function(expr, data, ignore_case, regex, verbose) {
+  first_obj <- .dynGet(expr[[2]], inherits = FALSE, minframe = 0L)
+  .eval_expr(
+    first_obj[[deparse(expr[[3]])]],
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+}
+
+# e.g ~ gear + cyl
+.select_tilde <- function(expr, data, ignore_case, regex, verbose) {
+  vars <- all.vars(expr)
+  unlist(lapply(
+    vars,
+    .eval_expr,
+    data = data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  ))
+}
+
+# e.g list(gear = 4, cyl = 5) [NOT SURE IT WILL BE USED]
+.select_list <- function(expr, data, ignore_case, regex, verbose) {
+  vars <- names(expr)
+  unlist(lapply(
+    vars,
+    .eval_expr,
+    data = data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  ))
+}
+
+# e.g is.numeric()
+.select_context <- function(expr, data, ignore_case, regex, verbose) {
+  x_dep <- insight::safe_deparse(expr)
+  if (endsWith(x_dep, "()")) {
+    new_expr <- gsub("\\(\\)$", "", x_dep)
+    new_expr <- str2lang(new_expr)
+    .eval_expr(
+      new_expr,
+      data = data,
+      ignore_case = ignore_case,
+      regex = regex,
+      verbose = verbose
+    )
+  } else {
+    .dynEval(expr, inherits = FALSE, minframe = 0L)
+  }
+}
+
+# -------------------------------------
+
+.check_data <- function(data) {
+  if (is.null(data)) {
+    insight::format_error("The `data` argument must be provided.")
+  }
+  .coerce_to_dataframe(data)
+}
+
+.regex_select_helper <- function() {
+  "(starts\\_with|ends\\_with|col\\_ends\\_with|contains|regex)"
+}
+
+.conflicting_packages <- function(packages = NULL) {
+  if (is.null(packages)) {
+    packages <- "poorman"
+  }
+
+  namespace <- vapply(packages, isNamespaceLoaded, FUN.VALUE = logical(1L))
+  attached <- paste0("package:", packages) %in% search()
+  attached <- stats::setNames(attached, packages)
+
+  for (i in packages) {
+    unloadNamespace(i)
+  }
+
+  list(package = packages, namespace = namespace, attached = attached)
+}
+
+.attach_packages <- function(packages = NULL) {
+  if (!is.null(packages)) {
+    pkg <- packages$package
+    for (i in seq_along(pkg)) {
+      if (isTRUE(packages$namespace[i])) {
+        loadNamespace(pkg[i])
+      }
+      if (isTRUE(packages$attached[i])) {
+        suppressPackageStartupMessages(
+          suppressWarnings(
+            require(pkg[i], quietly = TRUE, character.only = TRUE)
+          )
+        )
+      }
+    }
+  }
+}
+
+# Almost identical to dynGet(). The difference is that we deparse the expression
+# because get0() allows symbol only since R 4.1.0
+.dynGet <- function(x,
+                    ifnotfound = stop(gettextf("%s not found", sQuote(x)), domain = NA),
+                    minframe = 1L,
+                    inherits = FALSE) {
+  x <- insight::safe_deparse(x)
+  n <- sys.nframe()
+  myObj <- structure(list(.b = as.raw(7)), foo = 47L)
```

Comment:
What do the magic numbers `7L` and `47L` signify here?

```suggestion
  myObj <- structure(list(.b = as.raw(7)), foo = 47L)
```

Reply:
I don't think it has a meaning in itself, I just took the source code of `dynGet()` and added the `insight::safe_deparse()`